File: Python/StatePattern.py
# Context : AmazonOrder
class AmazonOrder:
    def __init__(self):
        self.state = PaidForState()

    def transition_to(self,state):
        print(f"Order transitioning from {type(self.state).__name__} to {type(state).__name__}")
        self.state = state

    def verify_payment(self):
        self.state.verify_payment(self)

    def ship_order(self):
        self.state.ship_order(self)

# State Interface
class OrderState():
    def verify_payment(self,order:AmazonOrder):
        pass
    def ship_order(self,order:AmazonOrder):
        pass

# Concrete States
class PaidForState(OrderState):
    def verify_payment(self,order:AmazonOrder):
        print("Payment verified.")
        order.transition_to(VerifiedPaymentState())

    def ship_order(self,order:AmazonOrder):
        print("Cannot ship order. Payment not yet verified")


class VerifiedPaymentState(OrderState):
    def verify_payment(self, order: AmazonOrder):
        print("Payment already verified.")

    def ship_order(self, order: AmazonOrder):
        print("Order shipped.")
        order.transition_to(ShippedState())

class ShippedState(OrderState):
    def verify_payment(self, order: AmazonOrder):
        print("Payment already verified.")

    def ship_order(self, order: AmazonOrder):
        print("Order already shipped.")


# Client code
if __name__ == "__main__":
    order = AmazonOrder()
    order.ship_order()   # Cannot ship before verification
    order.verify_payment()  # Verify payment
    order.ship_order()  # Ship the order
    order.verify_payment()  # Attempt to verify payment again

File: Typescript/StatePattern.js
// Context: AmazonOrder
var AmazonOrder = /** @class */ (function () {
    function AmazonOrder() {
        this.state = new PaidForState();
    }
    AmazonOrder.prototype.transitionTo = function (state) {
        console.log("Order transitioning from ".concat(this.state.getName(), " to ").concat(state.getName()));
        this.state = state;
    };
    AmazonOrder.prototype.verifyPayment = function () {
        this.state.verifyPayment(this);
    };
    AmazonOrder.prototype.shipOrder = function () {
        this.state.shipOrder(this);
    };
    return AmazonOrder;
}());
// Concrete States
var PaidForState = /** @class */ (function () {
    function PaidForState() {
    }
    PaidForState.prototype.getName = function () {
        return "PaidForState";
    };
    PaidForState.prototype.verifyPayment = function (order) {
        console.log("Payment verified.");
        order.transitionTo(new VerifiedPaymentState());
    };
    PaidForState.prototype.shipOrder = function (order) {
        console.log("Cannot ship order. Payment not yet verified.");
    };
    return PaidForState;
}());
var VerifiedPaymentState = /** @class */ (function () {
    function VerifiedPaymentState() {
    }
    VerifiedPaymentState.prototype.getName = function () {
        return "VerifiedPaymentState";
    };
    VerifiedPaymentState.prototype.verifyPayment = function (order) {
        console.log("Payment already verified.");
    };
    VerifiedPaymentState.prototype.shipOrder = function (order) {
        console.log("Order shipped.");
        order.transitionTo(new ShippedState());
    };
    return VerifiedPaymentState;
}());
var ShippedState = /** @class */ (function () {
    function ShippedState() {
    }
    ShippedState.prototype.getName = function () {
        return "ShippedState";
    };
    ShippedState.prototype.verifyPayment = function (order) {
        console.log("Payment already verified.");
    };
    ShippedState.prototype.shipOrder = function (order) {
        console.log("Order already shipped.");
    };
    return ShippedState;
}());
// Client code
var order = new AmazonOrder();
order.shipOrder(); // Cannot ship before verification
order.verifyPayment(); // Verify payment
order.shipOrder(); // Ship the order
order.verifyPayment(); // Attempt to verify payment again
